duracloud/duracloud
# Add code to use RabbitMQ, Swift, and SMTP email

# What does this Pull Request do?
We provide the implementation of the configuration changes made in duracloud-db and management-console.

We have added the ability to:
- Read audit logs from a Swift container
- Write audit events to a RabbitMQ queue
- Read/write notification events from/to a RabbitMQ queue
- Send emails using an SMTP server

# How should this be tested?

### Prerequisites for testing
- An SMTP server
- A RabbitMQ server with:
  - `rabbitmq_message_timestamp` plugin installed
  - a fanout exchange
  - a direct exchange
- A Swift endpoint running S3 emulation middleware layer

### Steps
You will need to build this repo first, then build management-console.

To test SMTP email, add this to the .properties file:
```
#Email Type: 'ses' or 'spring'
notification.type=spring
# Connection details for email notifications
notification.user=smtp_or_aws_username
notification.pass=smtp_or_aws_password
notification.from-address=someone@example.com
notification.admin-address=admin@example.com
#If the notification type is 'spring' (must be SMTP server)
notification.host=smtp.example.com
notification.port=25
```
then restart duradmin/durastore. The debug log should show a successful email connection test.

To test RabbitMQ notifications:
- Create a "fanout" exchange in RabbitMQ
- configure Global Properties in the AMA with RabbitMQ
- use the name of the fanout exchange as the "RabbitMQ exchange" option
- add a new space
- you should see successful message dispatch, receipt and deletion messages in the debug log
- you should also see that two queues were automatically created in RabbitMQ and bound to the fanout exchange

To test RabbitMQ audit queue:
- Create a "direct" exchange
- Create a queue and give it a {queuename}. Bind this queue to the direct exchange and use {queuename} as the routing key
- configure Duracloud Mill in the AMA with RabbitMQ
  - use the name of the direct exchange as the RabbitMQ Exchange
  - Use {queuename} as the Audit Queue
- Perform some audit-able actions. You should see messages appearing on the audit queue

To test audit logs:
- Add these parameters to the .properties file
```
swift.accessKey=accesskey
swift.secretKey=secretkey
swift.endpoint=https://s3.example.com
swift.signerType=signerType_if_required
```
- Restart duradmin/durastore
- Generate an audit log

# Notable changes
We have changed the package `package org.duracloud.common.sns` to `package org.duracloud.common.changenotifier` to more accurately reflect its purpose.

# Interested parties
@bbranan
